teps/0058-graceful-pipeline-run-termination.md

```diff
@@ -136,32 +149,59 @@ Consider including folks that also work outside the WGs or subproject.
-
-Consider including folks that also work on CLI and dashboard.
--->
+Support for additional command or flags in `tkn` CLI should be considered.
 
 ### Performance (optional)
 
 No impact on performance.
 
 ## Design Details
 
-<!--
-This section should contain enough information that the specifics of your
-change are understandable.  This may include API specs (though not always
-required) or even code snippets.  If there's any ambiguity about HOW your
-proposal will be implemented, this is the place to discuss them.
-
-If it's helpful to include workflow diagrams or any other related images,
-add them under "/teps/images/". It's upto the TEP author to choose the name
-of the file, but general guidance is to include at least TEP number in the
-file name, for example, "/teps/images/NNNN-workflow.jpg".
--->
+In this proposal we need to consider following cases:
+
+1. User sets `spec.status` to "PipelineRunStopped" in `PipelineRun` with running tasks, but `finally` section is empty.
+
+    In this case a graceful termination would ack the same as a pipeline run cancellation.
+    
+    - `spec.status` in all running `TaskRun` is patched to "TaskRunCancelled"
+    - `spec.status` in all running `Run` is patched to "RunCancelled"
+    - `PipelineRun` condition after graceful termination is:
+      ```yaml
+      { type: Succeeded, Status: False, Reason: Cancelled } 
+      ```
+
+2. User sets `spec.status` to "PipelineRunStopped" in `PipelineRun` with running tasks and non-empty `finally` section.
+
+    In this case a graceful termination cancels all running task runs and waits for final tasks to be processed.
+    
+    - `spec.status` in all running `TaskRun` is patched to "TaskRunCancelled"
+    - `spec.status` in all running `Run` is patched to "RunCancelled"
+    - `PipelineRun` condition after graceful termination is:
+      ```yaml
+      { type: Succeeded, Status: Unknown, Reason: PipelineRunStopping } 
+      ```    
+
+3. User sets `spec.status` to "PipelineRunStopping" in `PipelineRun` with running tasks.
+
+    In this case a graceful termination waits for all tasks to be completed.
+    
+    - `PipelineRun` condition after graceful termination is:
+      ```yaml
+      { type: Succeeded, Status: Unknown, Reason: PipelineRunStopping } 
+      ```    
+
+4. User sets `spec.status` to "PipelineRunStopped" or "PipelineRunStopping" in `PipelineRun` with running final tasks.
+
+    In this case a graceful termination does not change a pipeline run state, which waits for all final tasks to be completed.
+    
+    - `PipelineRun` condition is unchanged.
+
+5. User sets `spec.status` to "PipelineRunStopped" or "PipelineRunStopping" in `PipelineRun` with tasks not scheduled yet.
+
+    When a pipeline run is gracefully terminated (in the way described above), any unscheduled non-final task is skipped
+    and listed in `status.skippedTasks` in `PipelineRun`.
+    
+    Final tasks, if present, are scheduled normally.  
```

Comment:
This does not align with the `pipelineRunCancelled` but I see it as aligned with graceful termination. The tasks which are not scheduled are skipped and `pipelineRun` starts executing the `finally` tasks. 

`pipelineRunCancelled` cancels the running `taskRuns` and leaves remaining tasks as `incomplete`.

Reply:
@pritidesai the proposed behaviour is aligned with a task failure case.